File: src/types/metadata.ts
export interface ColumnMetadata {
  name: string;
  description: string;
}

export interface GeneratedMetadata {
  description: string;
  columns: ColumnMetadata[];
  exampleUsage: string;
  relatedObjects: string[];
  dependencies: string[];
}

export interface ObjectMetadata {
  id: string;
  connectionId: string;
  schemaName: string;
  objectName: string;
  objectType: "table" | "view" | "function" | "procedure";
  metadata: GeneratedMetadata;
  generatedAt: string;
  updatedAt: string;
}

export interface MetadataProgress {
  connectionId: string;
  current: number;
  total: number;
  objectName: string;
  status: "generating" | "complete" | "error";
}

export interface MetadataDone {
  connectionId: string;
  totalGenerated: number;
}

export interface MetadataError {
  connectionId: string;
  error: string;
}

File: src-tauri/src/metadata.rs
use serde::{Deserialize, Serialize};
use std::path::PathBuf;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ColumnMetadata {
    pub name: String,
    pub description: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GeneratedMetadata {
    pub description: String,
    pub columns: Vec<ColumnMetadata>,
    pub example_usage: String,
    pub related_objects: Vec<String>,
    pub dependencies: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ObjectMetadata {
    pub id: String,
    pub connection_id: String,
    pub schema_name: String,
    pub object_name: String,
    pub object_type: String, // "table", "view", "function", "procedure"
    pub metadata: GeneratedMetadata,
    pub generated_at: String,
    pub updated_at: String,
}

/// JSON file store for database object metadata.
pub struct MetadataStore {
    path: PathBuf,
}

impl MetadataStore {
    pub fn new(app_data_dir: &PathBuf) -> Self {
        Self {
            path: app_data_dir.join("metadata.json"),
        }
    }

    pub fn load(&self) -> Vec<ObjectMetadata> {
        match std::fs::read_to_string(&self.path) {
            Ok(data) => serde_json::from_str(&data).unwrap_or_default(),
            Err(_) => Vec::new(),
        }
    }

    pub fn save(&self, entries: &[ObjectMetadata]) -> Result<(), String> {
        let json = serde_json::to_string_pretty(entries)
            .map_err(|e| format!("Failed to serialize metadata: {e}"))?;
        std::fs::write(&self.path, json)
            .map_err(|e| format!("Failed to write metadata: {e}"))
    }
}
